### PR TITLE
Merge gc.h and internal/gc.h

### PR DESCRIPTION
**Pull request by eightbitraptor**
## History:

Currently we have 4 headers named `gc.h`

* `gc.h`
* `internal/gc.h`
* `include/ruby/internal/gc.h`
* `include/ruby/internal/intern/gc.h`

The first two are private and internal to the CRuby codebase, the latter 2 are visible to C extensions. This PR attempts to merge the two internal headers `gc.h` and internal/gc.h` into a single header `internal/gc.h` in order to simplify the codebase and put all the gc related interface in a single place.

Here is the history of the 2 internal only `gc.h` headers as I've been able to understand from the commit logs:

* `gc.h` was created as part of the YARV merge in 2006, it contained only some `defines` for debug related macros.
* In 2010 some function declarations were added to it to expose some functionality (`rb_objspace_each_objects`) to the `objspace` extension
* In May 2011 `internal.h` was created as a place to put shared, but internal only declarations.
* In June 2011 GC related functions started being committed to `internal.h`
* In 2019 `internal.h` was refactored and `internal/gc.h` was created

As of 2023 internally used code continues to be added to both `gc.h` and `internal/gc.h`, which one is chosen in any particular circumstance looks to be whichever header is already being included for legacy reasons in the `.c` file being edited.

### CODE REVIEW

File: compile.c [hunk 1]
```diff
@@ -17,9 +17,9 @@
 #endif
 
 #include "encindex.h"
-#include "gc.h"
 #include "id_table.h"
 #include "internal.h"
+#include "internal/gc.h"
```
Comment (peterzhu2118) [inline compile.c:22]: Can you fix all the `include` to be in alphabetical order?

File: misc/.vscode/settings.json [hunk 2]
```diff
@@ -2,4 +2,7 @@
     "rust-analyzer.cargo.features": [
         "disasm",
     ],
+    "files.associations": {
+        "id_table.h": "c"
+    },
```
Comment (peterzhu2118) [inline misc/.vscode/settings.json:7]: Why is this changed?

**Review by peterzhu2118 — COMMENTED** [review]
There's several places in `depend` files that have `$(top_srcdir)/gc.h`, I think these can also go away.

File: yjit/src/cruby_bindings.inc.rs [hunk 3]
```diff
@@ -1126,8 +1126,8 @@ extern "C" {
     pub fn rb_attr_get(obj: VALUE, name: ID) -> VALUE;
     pub fn rb_obj_info_dump(obj: VALUE);
     pub fn rb_reg_new_ary(ary: VALUE, options: ::std::os::raw::c_int) -> VALUE;
-    pub fn rb_class_allocate_instance(klass: VALUE) -> VALUE;
     pub fn rb_obj_info(obj: VALUE) -> *const ::std::os::raw::c_char;
+    pub fn rb_class_allocate_instance(klass: VALUE) -> VALUE;
     pub fn rb_shape_id_offset() -> i32;
```
Comment (peterzhu2118) [inline yjit/src/cruby_bindings.inc.rs:1131]: Why was this changed?

File: vm_sync.c [hunk 4]
```diff
@@ -2,7 +2,7 @@
 #include "vm_sync.h"
 #include "ractor_core.h"
 #include "vm_debug.h"
-#include "gc.h"
+#include "internal/gc.h"
```
Comment (peterzhu2118) [inline vm_sync.c:5]: Can you move this to the top so that it's in alphabetical order?

File: string.c [hunk 5]
```diff
@@ -23,7 +23,7 @@
 
 #include "debug_counter.h"
 #include "encindex.h"
-#include "gc.h"
+#include "internal/gc.h"
```
Comment (peterzhu2118) [inline string.c:26]: This one isn't in alphabetical order.

File: ractor.c [hunk 6]
```diff
@@ -14,7 +14,7 @@
 #include "internal/struct.h"
 #include "internal/thread.h"
 #include "variable.h"
-#include "gc.h"
+#include "internal/gc.h"
```
Comment (peterzhu2118) [inline ractor.c:17]: This one too.

File: mjit.c [hunk 7]
```diff
@@ -90,7 +90,7 @@
 #include "vm_callinfo.h"
 #include "mjit.h"
 #include "mjit_c.h"
-#include "gc.h"
+#include "internal/gc.h"
```
Comment (peterzhu2118) [inline mjit.c:93]: This one is out of order too.

File: ext/objspace/objspace.c [hunk 8]
```diff
@@ -12,8 +12,8 @@
 
 **********************************************************************/
 
-#include "gc.h"
 #include "internal.h"
+#include "internal/gc.h"
```
Comment (peterzhu2118) [inline ext/objspace/objspace.c:16]: This one too.

File: yjit/src/cruby_bindings.inc.rs (hunk 3 above)
Comment (eightbitraptor) [reply yjit/src/cruby_bindings.inc.rs:1131]: because the yjit github actions build complained at me and told me to apply the diff

File: mjit.c (hunk 7 above)
Comment (eightbitraptor) [reply mjit.c:93]: so is the entire of this list to be fair...

**Review by peterzhu2118 — APPROVED** [review]
